### ISSUE TITLE
Overload does not work correctly - the entire type signature system needs a rewrite

### ISSUE THREAD

**Issue by piotrmaslanka** [piotrmaslanka/satella#50]
Overload does not work correctly. Say:

```python
@overload
def what(x):
    print('Other')
@what.overload
def what(x: int):
    print('Integer')
```
and `what(None)` does not print Other.

This is going to require a complete rewrite of the signature system, and most probably the for_argument decorator as well

**Comment by piotrmaslanka** [comment]
Tagged <satella3> since this will require a major rewrite of how the typing system works.

**Comment by piotrmaslanka** [comment]
Closed as of #57
